### ISSUE TITLE
Self-contained .app: frozen sidecar, signing, notarization, first-run downloads

### ISSUE THREAD

**Issue by jacksodj** [jacksodj/academic-noosphere#22]
## Question

Make the downloaded `.app` a first-class Mac experience: double-click and it works — no uv, no repo checkout, no right-click-to-open. Two constraints are already decided by the map owner:

- **Signing/notarization is in**: an Apple Developer ID is available. Pipeline = Developer ID Application cert → codesign (hardened runtime) → `xcrun notarytool` → staple, wired into the Tauri build.
- **The SPECTER2 model (~440 MB) is NOT bundled**: the app downloads it post-install/first-run, with visible progress in onboarding.

Open decisions:

- **Freezing the Python core**: PyInstaller vs briefcase vs shipping a uv-managed private env inside app resources. The hard part is native deps (LadybugDB, DuckDB) and especially **torch** via sentence-transformers — a naively frozen sidecar could exceed 1–2 GB. Decide: freeze with torch, switch embedding to ONNX runtime, or make the whole embed stack a first-run download alongside the model.
- **Tauri integration**: use Tauri's `externalBin` sidecar mechanism (replaces the current `find_uv`/`repo_root` probing for the packaged path; keep `NOOSPHERE_REPO` as the dev-mode escape hatch).
- **First-run download UX**: what downloads (model only vs model + embed stack), from where, integrity checks, resumability, and where it lands (`~/Library/Application Support/academic-noosphere`).
- **Update story**: today `git pull` updates the core without a new download; a frozen app loses that. Decide the update channel (GitHub Releases + in-app check? Tauri updater plugin?) and cadence.
- **CI or local builds**: notarization needs the cert — decide whether releases stay local-Mac builds or move to CI with cert secrets.

Resolution = decisions recorded + a signed, notarized, self-contained release published (map carries execution).

Blocked by: (none)

Part of #21

**Comment by jacksodj** [comment]
## Direction: Option B (slim sealed app via ONNX), with two owner constraints (2026-08-24)

Options considered: **A** freeze everything incl. torch (1.5–2 GB app), **B** replace torch with ONNX Runtime (~150–250 MB app), **C** frozen core + hand-rolled runtime download of the torch stack (rejected: reimplements uv badly), **D** invisible-uv bootstrap (fallback if B stalls). **B chosen.**

**Constraint 1 — no local rebuild.** The existing graph (8k works, torch-produced SPECTER2 vectors) is untouched. The parity gate (embed a few hundred real abstracts with both stacks, compare cosines; fp32 export should agree to ~1e-5) is an **acceptance gate, not a migration trigger**: if it fails, the ONNX switch is blocked and we rethink — we do NOT re-embed the corpus. This also covers mixing: post-switch works embed via ONNX next to old torch vectors, which is only acceptable because the gate is strict.

**Constraint 2 — don't host what's already hosted elsewhere.** Checked the HF Hub (2026-08-24): `allenai/specter2_base` ships only `pytorch_model.bin` (439 MB) — **no official ONNX exists**; the only third-party ONNX is the wrong variant (`adamlabadorf/specter2-proximity-onnx`, 0 downloads) and is rejected on provenance. So the fp32 ONNX derivative we produce exists nowhere else → publishing it is not duplicative hosting. Plan: publish our verified export as a model repo on the **Hugging Face Hub** under the owner's account (models belong on HF, not in GitHub release assets), with the export script + parity report committed here for provenance. Tokenizer files (~1 MB) are pulled from `allenai/specter2_base` directly — those ARE hosted elsewhere and we don't copy them.

**Phasing (each step additive, shippable alone):**
1. Export script (dev-only, torch) → fp32 ONNX + recorded pooling recipe (mean-pool over attention mask + L2 normalize).
2. `OnnxSpecter2Embedder` (onnxruntime + tokenizers + numpy, ~50 MB) as a third implementation of the existing `Embedder` protocol in `src/noosphere/pipeline/embed.py` — no schema/API change; `default_embedder()` order: ONNX model present → sentence-transformers → stub, so dev workflow (`uv run` + torch extra) is unchanged and torch remains the parity oracle.
3. Parity gate run against the live corpus; results recorded here.
4. Publish export to HF; first-run onboarding downloads it (checksum-verified, progress UI) into `~/Library/Application Support`.
5. Only then: PyInstaller freeze of the torch-free core, Tauri `externalBin`, Developer ID signing + notarization + staple.

Open when the ticket is worked: parity threshold number, HF repo naming, whether first-run also offers the torch path on dev machines.

**Comment by jacksodj** [comment]
## Progress: phases 1–2 + packaging shipped (v0.2.0)

- **ONNX embedder landed** (`OnnxSpecter2Embedder`, `onnx` extra) behind the existing `Embedder` protocol; preference order ONNX → sentence-transformers → stub. Export via `scripts/export_specter2_onnx.py`.
- **Parity gate PASSED**: min cosine **0.99999982** over 110 texts (incl. 100 real OpenAlex abstracts) vs the torch oracle on MPS (the path that produced the corpus). Owner constraint honored: no re-embed, mixing safe. Report: `…/models/specter2-onnx/parity-report.json`.
- **Frozen sidecar**: PyInstaller onedir, 244 MB, torch-free (`packaging/noosphere-core.spec`), bundled at `Contents/Resources/sidecar`; shell prefers it, `NOOSPHERE_REPO` forces the dev/uv path.
- **Boot UX**: window opens instantly; core boots on a background thread, handshake injected late (`noosphere-ready`); SIGTERM now runs graceful exit (a TERM'd shell orphaned the core and held the DB lock during testing).
- **Signed**: Developer ID Application (Z3L3V842L2), hardened runtime + entitlements (`packaging/entitlements.plist`), every nested Mach-O signed (`packaging/sign-and-dmg.sh`). `codesign --verify --strict --deep` clean. Verified end-to-end: LaunchServices launch on this Mac runs the bundled core against the live corpus.
- **Released**: v0.2.0 with the signed DMG.

## Remaining before this ticket closes

1. **Notarization** — needs the one-time keychain profile (owner action):
   `xcrun notarytool store-credentials noosphere-notary --apple-id <apple-id> --team-id Z3L3V842L2` (app-specific password from appleid.apple.com), then re-run `packaging/sign-and-dmg.sh`.
2. **HF hosting of the ONNX artifact** — needs `hf auth login` once (owner action), then upload `models/specter2-onnx` + export provenance.
3. **First-run model download** in onboarding (checksum from `recipe.json`), pointing at the HF repo. Until then fresh machines fall back to the stub embedder (noted in the release).

**Comment by jacksodj** [comment]
## Resolution — shipped in v0.2.0

Option B delivered end to end; both owner constraints held:

- **ONNX path**: `OnnxSpecter2Embedder` behind the existing `Embedder` protocol; fp32 opset-17 export, **parity gate min cosine 0.99999982** vs the torch oracle over 110 texts (incl. 100 real OpenAlex abstracts). **No corpus re-embed** — the existing graph is untouched and mixing is provably safe.
- **Hosting**: no ONNX of `specter2_base` existed anywhere, so ours (the non-duplicative derivative) is published at [jacksodj/specter2-base-onnx](https://huggingface.co/jacksodj/specter2-base-onnx) with recipe, parity report, and provenance card; tokenizer pulled from the artifact, weights byte-faithful to allenai's Apache-2.0 model.
- **Frozen sidecar**: PyInstaller onedir (244 MB, torch excluded) at `Contents/Resources/sidecar`; shell prefers it, `NOOSPHERE_REPO` forces the uv dev path.
- **Signed + notarized + stapled**: Developer ID (Z3L3V842L2), hardened runtime, every nested Mach-O signed; Gatekeeper verdict `accepted — Notarized Developer ID`. `packaging/sign-and-dmg.sh` notarizes automatically via the `noosphere-notary` keychain profile.
- **First-run model download**: onboarding step + Settings panel; `ModelFetch` streams from HF with sha256 verification against `recipe.json` and atomic install. Fresh-machine simulation verified: stub → 418 MB download → onnx.
- **Boot UX**: instant window, background core boot, late handshake injection (`noosphere-ready`); SIGTERM runs graceful exit (a TERM'd shell had orphaned the core mid-session).
- **Update story (decided)**: GitHub Releases; the DMG is the update channel. In-app update checks deferred — fog on the map if wanted. **CI (decided)**: releases stay local-Mac builds; notarization credentials remain only in this Mac's keychain.

Released: [v0.2.0](https://github.com/jacksodj/academic-noosphere/releases/tag/v0.2.0) (notarized DMG, 118 MB). Contributor setup is now: download, drag, double-click, onboard.
